File: breadth-first/bf.js
/**
 * Breadth-first traversal.
 *
 * The first implementation does not contain checks for visited nodes.
 * Therefore, it will only work when there are no cycles in the graph,
 * i.e. trees.
 *
 * Sample tree:   0
 *               / \
 *             1     2
 *            / \   / \
 *           3   4 5   6
 *
 * The second implementation adds the aforementioned checks and becomes
 * a generic solution to any graph.
 *
 * Sample graph:    A---D
 *                 / \ /
 *                B---C--E---G
 *                    | /
 *                    |/
 *                    F
 */

"use strict";

const tree = {
    value: 0,
    children: [
        { value: 1, children: [ { value: 3, children: [] }, { value: 4, children: [] } ] },
        { value: 2, children: [ { value: 5, children: [] }, { value: 6, children: [] } ] }
    ]
};

function treeBreadthFirst(queue, visit) {
    while (queue.length > 0) {
        let node = queue.shift();
        visit(node);
        queue = queue.concat(node.children);
    }
}

function treeNodeVisitor(node) {
    treeNodeVisitor.values += node.value + " ";
}

treeNodeVisitor.values = "";
treeBreadthFirst([tree], treeNodeVisitor);

console.log("Tree breadth-first traversal:", treeNodeVisitor.values);

// -- Generic implementation allowing cycles

const graph1 = {
    "A": ["B", "C", "D"],
    "B": ["A", "C"],
    "C": ["A", "B", "D", "E", "F"],
    "D": ["A", "C"],
    "E": ["C", "F", "G"],
    "F": ["C", "E"],
    "G": ["E"]
};

function breadthFirst(graph, startingNode, visit) {
    const visited = [];
    let queue = [startingNode];
    while (queue.length != 0) {
        const node = queue.shift();
        if (visited.includes(node)) {
            continue;
        }
        visit(node);
        visited.push(node);
        queue = queue.concat(graph[node]);
    }
}

function graphNodeVisitor(node) {
    graphNodeVisitor.values += node + " ";
}

graphNodeVisitor.values = "";
breadthFirst(graph1, "A", graphNodeVisitor);

console.log("Graph breadth-first traversal:", graphNodeVisitor.values);

// --- Starting from a different node ('C')

graphNodeVisitor.values = "";
breadthFirst(graph1, "C", graphNodeVisitor);

console.log("Graph breadth-first traversal (starting from node 'C'):", graphNodeVisitor.values);

// --- Starting from a different node ('G')

graphNodeVisitor.values = "";
breadthFirst(graph1, "G", graphNodeVisitor);

console.log("Graph breadth-first traversal (starting from node 'G'):", graphNodeVisitor.values);

// --- Original tree represented by adjacency list

const graph2 = {
    0: [1, 2],
    1: [0, 3, 4],
    2: [0, 5, 6],
    3: [1],
    4: [1],
    5: [2],
    6: [2]
};

graphNodeVisitor.values = "";
breadthFirst(graph2, 0, graphNodeVisitor);

console.log("Graph breadth-first traversal (original tree):", graphNodeVisitor.values);

File: breadth-first/bf.go
package main

import "fmt"

type Publisher struct {
	name string
}

type Game struct {
	title     string
	year      int
	publisher *Publisher
}

type PublisherNode struct {
	publisher Publisher
}

type GameNode struct {
	game *Game
}

type Node struct {
	value string
	adj   []*Node
}

func createGraph() map[string]*Node {
	nA := Node{value: "A"}
	nB := Node{value: "B"}
	nC := Node{value: "C"}
	nD := Node{value: "D"}
	nE := Node{value: "E"}
	nF := Node{value: "F"}
	nG := Node{value: "G"}

	nA.adj = []*Node{&nB, &nC, &nD}
	nB.adj = []*Node{&nA, &nC}
	nC.adj = []*Node{&nA, &nB, &nD, &nE, &nF}
	nD.adj = []*Node{&nA, &nC}
	nE.adj = []*Node{&nC, &nF, &nG}
	nF.adj = []*Node{&nC, &nE}
	nG.adj = []*Node{&nE}

	graph := make(map[string]*Node)
	graph["A"] = &nA
	graph["B"] = &nB
	graph["C"] = &nC
	graph["D"] = &nD
	graph["E"] = &nE
	graph["F"] = &nF
	graph["G"] = &nG

	return graph
}

type VisitorFn func(*Node)

func present(node *Node, nodes []*Node) bool {
	for _, n := range nodes {
		if n == node {
			return true
		}
	}
	return false
}

func breadthFirst(startingNode *Node, visitorFn VisitorFn) {
	queue := []*Node{startingNode}
	var visited []*Node
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		if present(node, visited) {
			continue
		}
		visitorFn(node)
		visited = append(visited, node)
		queue = append(queue, node.adj...)
	}
}

func printNode(node *Node) {
	fmt.Printf("Visiting node '%s'\n", node.value)
}

func main() {
	graph := createGraph()
	fmt.Println("Breadth-first starting from 'A'")
	breadthFirst(graph["A"], printNode)
	fmt.Println("\nBreadth-first starting from 'C'")
	breadthFirst(graph["C"], printNode)
	fmt.Println("\nBreadth-first starting from 'G'")
	breadthFirst(graph["G"], printNode)
}
